jax_galsim/core/utils.py

```diff
@@ -13,54 +23,113 @@ def compute_major_minor_from_jacobian(jac):
+            dtype = float
+
     if isinstance(x, jax.Array):
-        return jnp.atleast_1d(x).astype(float).ravel()[0]
+        return jnp.atleast_1d(x).astype(dtype).ravel()[0]
     elif hasattr(x, "astype"):
-        return x.astype(float).ravel()[0]
+        return x.astype(dtype).ravel()[0]
     else:
-        return jnp.atleast_1d(jnp.array(x, dtype=float)).ravel()[0]
+        return jnp.atleast_1d(jnp.array(x, dtype=dtype)).ravel()[0]
 
 
 def cast_to_python_float(x):
     """Cast the input to a python float. Works on python floats and jax arrays.
     For jax arrays it always takes the first element after a call to .ravel()"""
     if isinstance(x, jax.Array):
-        return cast_to_float_array_scalar(x).item()
+        return _cast_to_array_scalar(x, dtype=float).item()
     else:
-        return float(x)
+        try:
+            return float(x)
+        except TypeError:
+            # this will return the same value for anything float-like that
+            # cannot be cast to float
+            # however, it will raise an error if something is not float-like
+            return 1.0 * x
+        except ValueError as e:
+            # we let NaNs through
+            if " NaN " in str(e):
+                # this will return the same value for anything float-like that
+                # cannot be cast to float
+                # however, it will raise an error if something is not float-like
+                return 1.0 * x
+            else:
+                raise e
 
 
-def cast_to_float(x):
-    """Cast the input to a float. Works on python floats and jax arrays."""
+def cast_to_python_int(x):
+    """Cast the input to a python int. Works on python ints and jax arrays.
+    For jax arrays it always takes the first element after a call to .ravel()"""
     if isinstance(x, jax.Array):
-        return x.astype(float)
-    elif hasattr(x, "astype"):
-        return x.astype(float)
+        return _cast_to_array_scalar(x, dtype=int).item()
     else:
         try:
-            return float(x)
-        except TypeError as e:
-            # needed so that tests of jax_galsim.angle pass
-            if "AngleUnit" in str(e):
+            return int(x)
+        except TypeError:
+            # this will return the same value for anything int-like that
+            # cannot be cast to int
+            # however, it will raise an error if something is not int-like
+            return 1 * x
+        except ValueError as e:
+            # we let NaNs through
+            if " NaN " in str(e):
+                # this will return the same value for anything int-like that
+                # cannot be cast to int
+                # however, it will raise an error if something is not int-like
+                return 1 * x
+            else:
                 raise e
 
-            return x
+
+def cast_to_float(x):
+    """Cast the input to a float. Works on python floats and jax arrays."""
+    try:
+        return float(x)
+    except Exception:
```

Comment:
The suggested change (without `Exception`) is bad practice actually. It would catch keyboard interrupts as well which would make programs not stop as expected. The code as written catches everything we want to catch.

Reply:
good to know!